homeassistant/components/apple_tv/config_flow.py

```diff
@@ -0,0 +1,403 @@
+            except exceptions.NoServiceError:
+                errors["base"] = "no_usable_service"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            # Use whatever the user entered as default value
+            default_suggestion = self.unique_id
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_IDENTIFIER, default=default_suggestion): str}
+            ),
+            errors=errors,
+            description_placeholders={"devices": self._devices_str()},
+        )
+
+    async def async_step_zeroconf(self, discovery_info):
+        """Handle device found via zeroconf."""
+
+        service_type = discovery_info[CONF_TYPE]
+        properties = discovery_info["properties"]
+
+        if service_type == "_mediaremotetv._tcp.local.":
+            identifier = properties["UniqueIdentifier"]
+            name = properties["Name"]
+        elif service_type == "_touch-able._tcp.local.":
+            identifier = discovery_info["name"].split(".")[0]
+            name = properties["CtlN"]
+        else:
+            return self.async_abort(reason="unrecoverable_error")
+
+        await self.async_set_unique_id(identifier, raise_on_progress=True)
+        self._abort_if_unique_id_configured(reload_on_update=False)
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context["identifier"] = self.unique_id
+        self.context["title_placeholders"] = {"name": name}
+        return await self.async_find_device_wrapper(self.async_step_confirm)
+
+    async def async_find_device_wrapper(self, next_func, allow_exist=False):
+        """Find a specific device and call another function when done.
+
+        This function will do error handling and bail out when an error
+        occurs.
+        """
+        try:
+            await self.async_find_device(allow_exist)
+        except DeviceNotFound:
+            return self.async_abort(reason="device_not_found")
+        except DeviceAlreadyConfigured:
+            return self.async_abort(reason="already_configured")
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            return self.async_abort(reason="unrecoverable_error")
+
+        return await next_func()
+
+    async def async_find_device(self, allow_exist=False):
+        """Scan for the selected device to discover services."""
+        self.scan_result, self.atv = await device_scan(
+            self.unique_id, self.hass.loop, cache=self.scan_result
+        )
+        if not self.atv:
+            raise DeviceNotFound()
+
+        self.protocol = self.atv.main_service().protocol
+
+        if not allow_exist:
+            for identifier in self.atv.all_identifiers:
```

Comment:
Does `self.atv.all_identifiers` only return not configured identifiers? Otherwise it seems weird to abort if already configured identifiers are used as unique ids, as that is expected.

Reply:
`self.atv.all_identifiers` returns all unique identifiers associated with the device. This handles this case:

User starts the flow and enters an identifier of a device that has already exists in HA. In this case I want to display an error and let the user enter something else.